File: lib/year2019/day08.rb
require 'solver'

module Year2019
  class Day08 < Solver
    def part_one
      image
        .layers
        .min_by { |layer| layer.count_of '0' }
        .count_lookup
        .slice('1', '2')
        .values
        .reduce(:*)
    end

    def part_two
      "\n" + image.display
    end

    def image
      @image ||= Image.new(raw_input.chomp, width, height)
    end

    def width
      25
    end

    def height
      6
    end

    class Image
      def initialize(string, width, height)
        @chars = string.chars
        @width = width
        @height = height
      end

      def top_layer
        Layer.new Array.new(@width * @height) { '2' }
      end

      def layers
        @layers ||= @chars.each_slice(@width * @height).map { |string| Layer.new string }
      end

      def decoded
        @decoded ||= layers.reduce(top_layer) { |upper, lower| upper.combine_with lower }
      end

      def display
        Grid.display(
          decoded.chars.each_slice(@width),
          :type => :rows,
          "0" => " ",
          "1" => "#",
        )
      end
    end

    class Layer
      attr_reader :chars

      def initialize(chars)
        @chars = chars
      end

      def combine_with(other)
        Layer.new @chars.zip(other.chars).map { |(upper, lower)| upper == '2' ? lower : upper }
      end

      def count_of(char)
        count_lookup[char]
      end

      def count_lookup
        @count_lookup ||= @chars.each_with_object(Hash.new { |h, k| h[k] = 0 }) do |char, hash|
          hash[char] += 1
        end
      end
    end
  end
end

File: rust/src/year2019/day08.rs
use crate::io::{Input, Solution};

pub fn solve(part: &str, input: &Input) -> Solution {
    Solution::build(part, input, &part_one, &part_two)
}

fn part_one(input: &Input) -> usize {
    validate(input, 25, 6)
}

fn part_two(input: &Input) -> usize {
    decode(input, 25, 6).show()
}

fn validate(input: &Input, width: usize, height: usize) -> usize {
    let layers = Layer::all(input, width, height);
    match layers
        .iter()
        .min_by(|l1, l2| l1.count('0').cmp(&l2.count('0')))
    {
        Some(layer) => layer.count('1') * layer.count('2'),
        None => 0,
    }
}

fn decode(input: &Input, width: usize, height: usize) -> Layer {
    let layers = Layer::all(input, width, height);
    layers
        .iter()
        .fold(Layer::transparent(width, height), |layer, lower| {
            layer.combine(lower)
        })
}

#[derive(Debug, Eq, PartialEq)]
struct Layer {
    digits: String,
    width: usize,
    height: usize,
}

impl Layer {
    fn all(input: &Input, width: usize, height: usize) -> Vec<Self> {
        let size = width * height;
        let mut layers = Vec::new();
        for i in 0..(input.data.len() / size) {
            let lower = i * size;
            let upper = lower + size;
            let digits = input.data[lower..upper].to_string();
            layers.push(Self {
                digits,
                width,
                height,
            });
        }
        layers
    }

    fn count(&self, digit: char) -> usize {
        self.digits
            .chars()
            .fold(0, |acc, ch| if ch == digit { acc + 1 } else { acc })
    }

    fn combine(&self, lower: &Layer) -> Self {
        let width = self.width;
        let height = self.height;
        let digits = (0..self.digits.len())
            .map(|i| {
                if self.char_at(i) == '2' {
                    lower.char_at(i)
                } else {
                    self.char_at(i)
                }
            })
            .collect::<String>();
        Self {
            digits,
            width,
            height,
        }
    }

    fn char_at(&self, index: usize) -> char {
        self.digits.as_bytes()[index] as char
    }

    fn show(&self) -> usize {
        for i in 0..(self.digits.len() / self.width) {
            let lower = i * self.width;
            let upper = lower + self.width;
            let display = self.digits[lower..upper]
                .chars()
                .map(|ch| if ch == '1' { '#' } else { ' ' })
                .collect::<String>();
            println!("{}", display);
        }
        1
    }

    fn transparent(width: usize, height: usize) -> Self {
        let size = width * height;
        let digits = (0..size).map(|_| '2').collect::<String>();
        Self {
            digits,
            width,
            height,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn input1() -> Input {
        Input::from("123456789012")
    }

    fn input2() -> Input {
        Input::from("0222112222120000")
    }

    #[test]
    fn test_part_one() {
        assert_eq!(validate(&input1(), 3, 2), 1);
    }

    #[test]
    fn test_part_two() {
        assert_eq!(
            decode(&input2(), 2, 2),
            Layer {
                digits: "0110".to_string(),
                width: 2,
                height: 2
            }
        );
    }
}
